cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/DeleteAndCreateSchedulesStage.java

```diff
@@ -42,25 +44,27 @@ public DeleteAndCreateSchedulesStage(Scheduler programScheduler) {
 
   @Override
   public void process(final ApplicationWithPrograms input) throws Exception {
-
-    if (!input.canUpdateSchedules()) {
-      // if we cant update schedules, emit and return
-      emit(input);
-      return;
-    }
-
     ApplicationId appId = input.getApplicationId();
+    Map<String, WorkflowSpecification> workflows = input.getSpecification().getWorkflows();
     // Get a set of new schedules from the app spec
     Set<ProgramSchedule> newSchedules = getProgramScheduleSet(appId, input.getSpecification());
     for (ProgramSchedule schedule : programScheduler.listSchedules(appId)) {
       if (newSchedules.contains(schedule)) {
         newSchedules.remove(schedule); // Remove the existing schedule from the newSchedules
         continue;
       }
-      // Delete the existing schedule if it is not present in newSchedules
-      programScheduler.deleteSchedule(schedule.getScheduleId());
+      String workflowName = schedule.getProgramId().getProgram();
+      // Delete the existing schedule if it is not present in newSchedules or if the workflow is deleted
+      if (input.canUpdateSchedules() || !workflows.containsKey(workflowName))  {
```

Comment:
@albertshau adding a check here to make sure that the workflow exists for the schedule

Reply:
I'm still not understanding how turning on the feature flag is breaking the existing logic. Does it change what `programScheduler.listSchedules()` returns for some reason?

Reply:
`programScheduler.listSchedules()` iterates over all workflows of the app to return the schedules. When FF is on, there are multiple workflows -one corresponding to each version. We check for a previous version workflow in `listSchedules` https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java#L551.  When FF is off there is only one workflow corresponding to `-SNAPSHOT`, which is deleted in the current app spec - as such the schedules are not returned.

Reply:
I took a look and `programScheduler.listSchedules()` just does a scan with `getScheduleKeysForApplicationScan` as the key, where the key is namespace, app name, and hardcoded `-snapshot` as the version. It does not look like there is any difference in the list logic between FF on or off.

This means the *write* logic is different. There is probably an existing bug where we are incorrectly using the application version when writing when it is supposed to always be `-snapshot`

Reply:
Ok looked into it a bit more. For the failing test case, this loop is never entered regardless of the feature flag value.

Instead, the buggy part is in the DeletedProgramHandlerStage. With the flag off, it sees things that programs were deleted and deletes the corresponding schedules. With the flag on, it doesn't seen any deleted programs. We should be fixing that part of the code, not this part.